indexing-hadoop/src/main/java/io/druid/indexer/InputRowHelper.java

```diff
@@ -0,0 +1,288 @@
+import org.apache.hadoop.io.ArrayWritable;
+import org.apache.hadoop.io.BytesWritable;
+import org.apache.hadoop.io.FloatWritable;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.MapWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.io.Writable;
+import org.joda.time.DateTime;
+
+import java.io.DataInput;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ */
+public class InputRowHelper
+{
+  private static final Logger log = new Logger(InputRowHelper.class);
+
+  private static final Text[] EMPTY_TEXT_ARRAY = new Text[0];
+
+  private final AggregatorFactory[] aggs;
+
+  public InputRowHelper(AggregatorFactory[] aggregators)
+  {
+    this.aggs = aggregators;
+  }
+
+  public final byte[] toBytes(final InputRow row)
+  {
+    try {
+      ByteArrayDataOutput out = ByteStreams.newDataOutput();
+
+      List<String> dimList = row.getDimensions();
+
+      Text[] dims = EMPTY_TEXT_ARRAY;
+      if(dimList != null) {
+        dims = new Text[dimList.size()];
+        for (int i = 0; i < dims.length; i++) {
+          dims[i] = new Text(dimList.get(i));
+        }
+      }
+      StringArrayWritable sw = new StringArrayWritable(dims);
+      sw.write(out);
+
+      out.writeLong(row.getTimestampFromEpoch());
+
+      MapWritable mw = new MapWritable();
+
+      //writing all dimensions
+      if(dimList != null) {
+        for (String dim : dimList) {
+          List<String> dimValue = row.getDimension(dim);
+
+          if (dimValue == null || dimValue.size() == 0) {
+            continue;
+          }
+
+          if (dimValue.size() == 1) {
+            mw.put(new Text(dim), new Text(dimValue.get(0)));
+          } else {
+            Text[] dimValueArr = new Text[dimValue.size()];
+            for (int i = 0; i < dimValueArr.length; i++) {
+              dimValueArr[i] = new Text(dimValue.get(i));
+            }
+            mw.put(new Text(dim), new StringArrayWritable(dimValueArr));
+          }
+        }
+      }
+
+      //writing all metrics
+      Supplier<InputRow> supplier = new Supplier<InputRow>()
+      {
+        @Override
+        public InputRow get()
+        {
+          return row;
+        }
+      };
+      for (AggregatorFactory aggFactory : aggs) {
+        String k = aggFactory.getName();
+
+        Aggregator agg = aggFactory.factorize(
+            IncrementalIndex.makeColumnSelectorFactory(
+                aggFactory,
+                supplier,
+                true
+            )
+        );
+        agg.aggregate();
+
+        String t = aggFactory.getTypeName();
+
+        if (t.equals("float")) {
+          mw.put(new Text(k), new FloatWritable(agg.getFloat()));
+        } else if (t.equals("long")) {
+          mw.put(new Text(k), new LongWritable(agg.getLong()));
+        } else {
+          //its a complex metric
+          Object sketch = agg.get();
```

Comment:
I am just calling the "intermediate representation of aggregate" a sketch, it can be anything.. no assumptions that it has to be something like hll.